backend/src/rendering.py

```diff
@@ -11,17 +11,76 @@
     © 2022 BeyondRGB
     This code is licensed under the MIT license (see LICENSE.txt for details)
 """
+# Python imports
+import numpy as np
+import gc
+
+# Local imports
+from constants import PROPHOTO_TRANS, COLORSPACE
 
 
 def render(packet):
     """ Render image pair as a single color calibrated image
-    [in]     imgs   : tuple of images to render
-    [in]     M      : calibration ndarray
+    [in] packet : pipeline packet
     [out] rendered image
     [post] rendered image is loaded in memory
+    [post] camsigs deleted from packet
+    """
+    camsigs = __genimgsigs(packet)
+
+    # Compute color calibrated image
+    m = np.resize(packet.x[0:18], (3, 6))
+    o = np.resize(packet.x[18:], (6, 1))
+    xyz = np.matmul(m, np.subtract(camsigs, o))
+    del camsigs
+    gc.collect()
+
+    # Convert to ProPhoto color space
+    if COLORSPACE == 'ProPhoto':
+        rgb = np.matmul(PROPHOTO_TRANS, xyz)
+        del xyz
+        gc.collect()
+        np.clip(rgb, 0, 1, out=rgb)
+        np.piecewise(rgb, [rgb > 0.001953125, rgb <= 0.001953125],
+                     [lambda rgb: rgb ** (1/1.8), lambda rgb: rgb * 16])
+
+    # Convert to sRGB color space
+    if COLORSPACE == 'sRGB':
+        rgb = np.matmul(PROPHOTO_TRANS, xyz)
+        del xyz
+        gc.collect()
+        np.clip(rgb/100, 0, 1, out=rgb)  # TODO remove divide by 100
+        np.piecewise(rgb, [rgb > 0.0031308, rgb <= 0.0031308],
```

Comment:
Reminder that we probably want to extract these into constants

Reply:
I made the cutoffs constants. Was not sure what to name the rest so I left them as is. The issue is the gamma numbers are provided without comment in the slides.

Reply:
Feel free to reopen if this is not satisfactory